testtools/matchers/_dict.py

```diff
@@ -235,15 +237,15 @@ class KeysEqual(Matcher):
     def __init__(self, *expected):
         """Create a `KeysEqual` Matcher.
 
-        :param expected: The keys the dict is expected to have.  If a dict,
-            then we use the keys of that dict, if a collection, we assume it
-            is a collection of expected keys.
+        :param expected: The keys the matchee is expected to have. As a
+            special case, if a single argument is specified, and it is a
+            mapping, then we use its keys as the expected set.
         """
         super(KeysEqual, self).__init__()
-        try:
+        if len(expected) == 1 and isinstance(expected[0], Mapping):
```

Comment:
It seems there's a slight change in behavior too here, the current code was happy with anything implementing a keys() method returning a list. With this branch we'll also require it to be a subclass of Mapping, I think. Perhaps something like:

```python

if len(expected) == and safe_hasattr(expected[0], "keys"):

would be a bit better, from a backward-compatibility point of view.